subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptEvaluator.kt

```diff
@@ -0,0 +1,254 @@
+package org.gradle.kotlin.dsl.provider
+
+import org.gradle.api.Project
+import org.gradle.api.initialization.dsl.ScriptHandler
+import org.gradle.api.internal.initialization.ClassLoaderScope
+import org.gradle.api.internal.initialization.ScriptHandlerInternal
+import org.gradle.api.internal.plugins.PluginAwareInternal
+
+import org.gradle.cache.CacheOpenException
+import org.gradle.cache.PersistentCache
+import org.gradle.cache.internal.CacheKeyBuilder
+
+import org.gradle.groovy.scripts.ScriptSource
+import org.gradle.groovy.scripts.internal.ScriptSourceHasher
+
+import org.gradle.internal.classpath.ClassPath
+import org.gradle.internal.classpath.DefaultClassPath
+import org.gradle.internal.hash.HashCode
+
+import org.gradle.kotlin.dsl.cache.ScriptCache
+
+import org.gradle.kotlin.dsl.execution.EvalOption
+import org.gradle.kotlin.dsl.execution.EvalOptions
+import org.gradle.kotlin.dsl.execution.Interpreter
+
+import org.gradle.kotlin.dsl.get
+
+import org.gradle.kotlin.dsl.support.EmbeddedKotlinProvider
+import org.gradle.kotlin.dsl.support.ImplicitImports
+import org.gradle.kotlin.dsl.support.KotlinScriptHost
+import org.gradle.kotlin.dsl.support.ScriptCompilationException
+import org.gradle.kotlin.dsl.support.transitiveClosureOf
+
+import org.gradle.plugin.management.internal.DefaultPluginRequests
+import org.gradle.plugin.management.internal.PluginRequests
+
+import java.io.File
+
+
+interface KotlinScriptEvaluator {
+
+    fun evaluate(
+        target: Any,
+        scriptSource: ScriptSource,
+        scriptHandler: ScriptHandler,
+        targetScope: ClassLoaderScope,
+        baseScope: ClassLoaderScope,
+        topLevelScript: Boolean,
+        options: EvalOptions
+    )
+}
+
+
+internal
+class StandardKotlinScriptEvaluator(
+    private val classPathProvider: KotlinScriptClassPathProvider,
+    private val classloadingCache: KotlinScriptClassloadingCache,
+    private val pluginRequestsHandler: PluginRequestsHandler,
+    private val embeddedKotlinProvider: EmbeddedKotlinProvider,
+    private val classPathModeExceptionCollector: ClassPathModeExceptionCollector,
+    private val kotlinScriptBasePluginsApplicator: KotlinScriptBasePluginsApplicator,
+    private val scriptSourceHasher: ScriptSourceHasher,
+    private val scriptCache: ScriptCache,
+    private val implicitImports: ImplicitImports
+) : KotlinScriptEvaluator {
+
+    override fun evaluate(
+        target: Any,
+        scriptSource: ScriptSource,
+        scriptHandler: ScriptHandler,
+        targetScope: ClassLoaderScope,
+        baseScope: ClassLoaderScope,
+        topLevelScript: Boolean,
+        options: EvalOptions
+    ) {
+
+        // TODO:partial-evaluator Optimise away this call whenever possible via partial evaluation
+        setupEmbeddedKotlinForBuildscript(scriptHandler)
+
+        val sourceHash = scriptSourceHasher.hash(scriptSource)
+
+        if (EvalOption.IgnoreErrors in options)
+            classPathModeExceptionCollector.ignoringErrors {
```

Comment:
:x: Extract condition and error collection in its own function in order not to repeat the `eval()` call